tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInRuleWithArrays.scala

```diff
@@ -0,0 +1,83 @@
+package at.forsyte.apalache.tla.bmcmt.rules
+
+import at.forsyte.apalache.tla.bmcmt.implicitConversions.Crossable
+import at.forsyte.apalache.tla.bmcmt.rewriter.ConstSimplifierForSmt
+import at.forsyte.apalache.tla.bmcmt.types.{BoolT, CellT, FinSetT, PowSetT}
+import at.forsyte.apalache.tla.bmcmt.{ArenaCell, SymbState, SymbStateRewriter, types}
+import at.forsyte.apalache.tla.lir.TlaEx
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir.UntypedPredefs._
+
+class SetInRuleWithArrays(rewriter: SymbStateRewriter) extends SetInRule(rewriter) {
+  private val simplifier = new ConstSimplifierForSmt
+
+  override protected def powSetIn(state: SymbState, powsetCell: ArenaCell, elemCell: ArenaCell): SymbState = {
+    def checkType: PartialFunction[(CellT, CellT), Unit] = {
+      case (PowSetT(FinSetT(expectedType)), FinSetT(actualType)) =>
+        assert(expectedType == actualType)
+    }
+    // double check that the type finder did its job
+    checkType(powsetCell.cellType, elemCell.cellType)
+
+    val leftElems = state.arena.getHas(elemCell)
+    val rightDomain = state.arena.getDom(powsetCell)
+    val rightDomainElems = state.arena.getHas(rightDomain)
```

Comment:
I find the names "left" and "right" confusing, especially that it is in the order opposite to the order of arguments. Shall we call them `setDomain`, `setDomainElems`, and`powsetElems`?

Reply:
I meant "left" and "right" here as the order in the `\in` operator. I think you suggestion does make things more clear. Thanks!